keylime/keylime_agent.py

```diff
@@ -620,74 +686,33 @@ def main():
-            action_list_path = os.path.join(secdir, "unzipped/action_list")
-            if os.path.exists(action_list_path):
-                with open(action_list_path, encoding="utf-8") as f:
-                    actionlisttxt = f.read()
-                if actionlisttxt.strip() != "":
-                    localactions = actionlisttxt.strip().split(',')
-                    for action in localactions:
-                        if not action.startswith('local_action_'):
-                            logger.warning("Invalid local action: %s. Must start with local_action_", action)
-                        else:
-                            actionlist.append(action)
-
-                    uzpath = "%s/unzipped" % secdir
-                    if uzpath not in sys.path:
-                        sys.path.append(uzpath)
-
-            for action in actionlist:
-                logger.info("Executing revocation action %s", action)
-                try:
-                    module = importlib.import_module(action)
-                    execute = getattr(module, 'execute')
-                    asyncio.get_event_loop().run_until_complete(execute(revocation))
-                except Exception as e:
-                    logger.warning("Exception during execution of revocation action %s: %s", action, e)
-        try:
-            while True:
-                try:
-                    revocation_notifier.await_notifications(
-                        perform_actions, revocation_cert_path=cert_path)
-                except Exception as e:
-                    logger.exception(e)
-                    logger.warning("No connection to revocation server, retrying in 10s...")
-                    time.sleep(10)
-        except KeyboardInterrupt:
-            logger.info("TERM Signal received, shutting down...")
-            instance_tpm.flush_keys()
-            server.shutdown()
-    else:
-        try:
-            while True:
-                time.sleep(1)
-        except KeyboardInterrupt:
-            logger.info("TERM Signal received, shutting down...")
-            instance_tpm.flush_keys()
-            server.shutdown()
+    def shutdown_handler(*_):
+        logger.info("TERM Signal received, shutting down...")
+        logger.debug("Stopping revocation notifier...")
+        revocation_process.terminate()
+        logger.debug("Shutting down HTTP server...")
+        server.shutdown()
+        server.server_close()
+        serverthread.join()
+        logger.debug("...HTTP server stopped")
+        revocation_process.join()
+        logger.debug("... revocation notifier stopped")
+        instance_tpm.flush_keys()
+        logger.debug("Flushed keys successfully")
+        sys.exit(0)
+
+    signal.signal(signal.SIGTERM, shutdown_handler)
+    signal.signal(signal.SIGQUIT, shutdown_handler)
+    signal.signal(signal.SIGINT, shutdown_handler)
+
+    # Keep the main thread alive by waiting for the server thread
+    serverthread.join()
```

Comment:
do we need to do `revocation_process.join()`?

maybe we can register the payload script thread too, and join it here too?

Reply:
Not really because this will already keep the main thread active until we call the shutdown handler.